crates/blockless-drivers/src/cgi_driver/db.rs

```diff
@@ -0,0 +1,185 @@
+
+#[derive(Clone, Copy)]
+pub(crate) enum ExtensionMetaStatus {
+    Normal = 0,
+    UPDATE = 1,
+    Invalid = -1,
+}
+
+impl Default for ExtensionMetaStatus {
+    fn default() -> Self {
+        ExtensionMetaStatus::Normal
+    }
+}
+
+impl From<i32> for ExtensionMetaStatus {
+    fn from(value: i32) -> Self {
+        match value {
+            0 => ExtensionMetaStatus::Normal,
+            -1 | _ => ExtensionMetaStatus::Normal,
+        }
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct ExtensionMeta {
+    pub id: i32,
+    pub md5: String,
+    pub description: String,
+    pub alias: String,
+    pub file_name: String,
+    pub status: ExtensionMetaStatus,
+}
+
+pub(crate) struct DB {
+    connect: Connection,
+}
+
+impl DB {
+    pub(crate) fn new(file: impl AsRef<Path>) -> Result<DB> {
+        let connect = Connection::open(file)?;
+        Ok(Self { connect })
+    }
+
+    pub(crate) fn create_schema(&mut self) -> Result<()> {
+        let schema_sql = r#"
+            create table if not exists extension_meta (
+                id INTEGER PRIMARY KEY,
+                alias TEXT NOT NULL,
+                md5 TEXT NOT NULL,
+                file_name TEXT NOT NULL,
+                status INTEGER DEFAULT 0,
+                description TEXT NOT NULL
+            );
+            UPDATE sqlite_sequence SET seq = 0 WHERE name = 'extension_meta';
+        "#;
+        self.connect.execute(schema_sql, ())?;
+        Ok(())
+    }
+
+    pub(crate) fn list_extensions(&self) -> Result<Vec<ExtensionMeta>> {
+        let query_sql = r#"
+            select id, alias, md5, file_name, description, status
+            from extension_meta where status = 0;
+        "#;
+        let mut stmt = self.connect.prepare(query_sql)?;
+        Ok(stmt
+            .query_map([], |row| {
+                let id = row.get(0)?;
+                let alias = row.get(1)?;
+                let md5 = row.get(2)?;
+                let file_name = row.get(3)?;
+                let description = row.get(4)?;
+                let status = row.get::<usize, i32>(5)?.into();
+                Ok(ExtensionMeta {
+                    id,
+                    md5,
+                    alias,
+                    status,
+                    file_name,
+                    description,
+                })
+            })
+            .map(|rows| rows.filter_map(|row| row.ok()).collect::<Vec<_>>())?)
+    }
+
+    pub(crate) fn save_extensions(&mut self, exts: &Vec<ExtensionMeta>) -> Result<()> {
+        for meta in exts.iter() {
+            match meta.status {
+                ExtensionMetaStatus::Normal => self.insert_extension_meta(meta),
+                ExtensionMetaStatus::UPDATE => self.update_extension_meta(meta),
+                ExtensionMetaStatus::Invalid => self.delete_extension_meta(meta),
+            }?;
+        }
+        Ok(())
+    }
+
+    pub(crate) fn delete_extension_meta(&mut self, meta: &ExtensionMeta) -> Result<()> {
```

Comment:
Also consider adding 'file_name' in the delete_extension_meta function, since 'id' alone might not be enough for proper deletion.

同时考虑在 delete_extension_meta 函数中添加 'file_name'，因为仅使用 'id' 可能不够。

Reply:
that's good idea